### ISSUE TITLE
Apple Silicon (`arm64e`) support

### ISSUE THREAD

**Issue by wimleers** [acidanthera/bugtracker#1928]
[NVMeFix 1.0.9](https://github.com/acidanthera/NVMeFix/releases/tag/1.0.9) and [Lilu 1.5.9](https://github.com/acidanthera/Lilu/releases/tag/1.5.9) are not compiled for Apple Silicon aka `arm64e` support:

```
lipo -info NVMeFix-1.0.9-RELEASE/NVMeFix.kext/Contents/MacOS/NVMeFix
Non-fat file: NVMeFix-1.0.9-RELEASE/NVMeFix.kext/Contents/MacOS/NVMeFix is architecture: x86_64
```
```
lipo -info Lilu-1.5.8-RELEASE/Lilu.kext/Contents/MacOS/Lilu 
Architectures in the fat file: Lilu-1.5.8-RELEASE/Lilu.kext/Contents/MacOS/Lilu are: i386 x86_64 
```

## But some Acidanthera  projects *are* available for Apple Silicon

In contrast with [`gfxutil 1.82b`](https://github.com/acidanthera/gfxutil/releases/tag/1.82b) (released 5 days ago, which btw oddly links to/has `1.81b` as file names for its attached downloads, not `1.82b` — I first thought I made a mistake 😅 ):

```
lipo -info gfxutil-1.81b-RELEASE/gfxutil
Architectures in the fat file: gfxutil-1.81b-RELEASE/gfxutil are: x86_64 arm64 
```

versus `1.80b`:
```
lipo -info gfxutil-1.80b-RELEASE/gfxutil
Architectures in the fat file: gfxutil-1.81b-RELEASE/gfxutil are: x86_64 i386
```

## Recompiling for `arm64e`

I figured I'd just build this for `arm64e` (so, change the [target architecture](https://therahulgoel.medium.com/xcode-12-upgrade-apple-silicon-and-arm64-a35ff6bfebed)), but … this does not seem to be trivial for those who haven't built code for macOS in a decade — when attempting to build Lilu I get errors about missing header files.

Any chance you could compile it for Apple silicon too? 🙏

**Comment by TheHackGuy** [comment]
why should you use lilu and nvmefix on apple silicon? 

Those missing headers is probably because you don't have mackernelsdk in the project yet.

Building Lilu
- Clone lilu from github
- open terminal and go to the Lilu project directory
- clone MacKernelSDK (git clone https://github.com/acidanthera/MacKernelSDK)
- change settings in xcode to arm64
- Build Lilu

Building NVMeFix
- Clone NVMeFix from github
-  open terminal and go to the NVMeFix project directory
- clone MacKernelSDK (git clone https://github.com/acidanthera/MacKernelSDK)
- - Copy the builded Lilu kext for arm to the NVMeFix project diectory
- change settings in xcode to arm64
- Build NVMeFix

I don't expect Lilu to work directly on arm without code changes.

**Comment by wimleers** [comment]
> Why should you use lilu and nvmefix on apple silicon?

**tl;dr**: To be able to use NVMeFix to enable APST on NVMEs, which macOS does not support natively.

More context? See below.

I have a M1 Mac Mini as a very energy-efficient home server. It idles at 4 Watt! (For back-ups, homebridge.io and software development from my iPad.)

I added an NVME drive via Thunderbolt. This consumes an additional 4 Watt! 🤯🙃 Even though I picked an energy-efficient one.

Because: macOS won’t make it enter its lower power states. That is where NVMeFix would help me 🤓

So: this would cut my home server’s energy use in half!

**Comment by vit9696** [comment]
Making Lilu/NVMeFix support Apple Silicon is between very difficult and impossible, since Apple Silicon uses authenticated pointers and hardware W^X, which prevents any dynamic memory patching in the kernel. I am afraid an entirely separate project needs to be created to support your use-case.

**Comment by wimleers** [comment]
Thank you so much for the incredibly fast response!

Amazing project 👏

**Comment by mdbraber** [comment]
Checking in with a related question @vit9696: is this also true for other projects such as AMFIExemption? Would it be possible to compile those for arm64e?

**Comment by Baekalfen** [comment]
> I added an NVME drive via Thunderbolt. This consumes an additional 4 Watt! 🤯🙃 Even though I picked an energy-efficient one.
> 
> Because: macOS won’t make it enter its lower power states. That is where NVMeFix would help me 🤓

A bit of self-promotion on my part, but I have a [user-space NVMe driver called MacVFN](https://github.com/samsungds/macvfn?tab=readme-ov-file#todo-list) to you might find interesting.

Problem is, that it doesn't register as a [BlockStorageDevice](https://developer.apple.com/documentation/blockstoragedevicedriverkit) yet, but if you're up for the task, I'll gladly help you implement it.

**Comment by vit9696** [comment]
To compile this to arm64e Lilu needs to be ported first.
